src/test/scala/temple/DSL/semantics/ValidatorTest.scala

```diff
@@ -14,210 +14,230 @@ class ValidatorTest extends FlatSpec with Matchers {
+          "User" -> ServiceBlock(
+            attributes = Map("a" -> Attribute(IntType())),
+            metadata = Seq(Readable.All, Readable.This),
+          ),
+        ),
+      ),
+    ) shouldBe Set("Multiple occurrences of Readable metadata in User")
+
+    validationErrors(
+      Templefile(
+        "MyProject",
+        services = Map(
+          "User" -> ServiceBlock(
+            attributes = Map("a" -> Attribute(IntType())),
+            metadata = Seq(Writable.All, Readable.This),
+          ),
+        ),
+      ),
+    ) shouldBe Set("#writable(all) is not compatible with #readable(this) in User")
+  }
 
-    the[SemanticParsingException] thrownBy {
+  it should "validate" in {
+    noException shouldBe thrownBy {
       validate(
         Templefile(
           "MyProject",
-          projectBlock = ProjectBlock(Seq(Database.Postgres)),
+          projectBlock = ProjectBlock(Seq(Database.Postgres, Provider.AWS)),
           services = Map(
             "User" -> ServiceBlock(
-              attributes = Map(
-                "a" -> Attribute(IntType()),
-              ),
-              metadata = Seq(Metadata.Uses(Seq("Box"))),
+              attributes = Map("a" -> Attribute(IntType())),
+              metadata = Seq(ServiceAuth.Email),
             ),
           ),
         ),
       )
-    } should have message singleError("No such service Box referenced in #uses in User")
+    }
 
     the[SemanticParsingException] thrownBy {
       validate(
         Templefile(
           "MyProject",
+          projectBlock = ProjectBlock(Seq(Database.Postgres)),
           services = Map(
             "User" -> ServiceBlock(
-              attributes = Map("a" -> Attribute(IntType())),
-              metadata = Seq(Readable.All, Readable.This),
+              attributes = Map(
+                "a" -> Attribute(IntType()),
+              ),
+              structs = Map(
+                "User" -> StructBlock(Map()),
+              ),
             ),
           ),
         ),
       )
-    } should have message singleError("Multiple occurrences of Readable metadata in User")
+    } should have message {
+      """An error was encountered while validating the Templefile
+        |Project, targets and structs must be globally unique, duplicate found: User (service, struct)""".stripMargin
+    }
 
     the[SemanticParsingException] thrownBy {
       validate(
-        Templefile(
-          "MyProject",
-          services = Map(
-            "User" -> ServiceBlock(
-              attributes = Map("a" -> Attribute(IntType())),
-              metadata = Seq(Writable.All, Readable.This),
-            ),
-          ),
+        templefileWithUserAttributes(
+          "b" -> Attribute(FloatType(max = Some(0), min = Some(1))),
+          "c" -> Attribute(FloatType(precision = -1)),
         ),
       )
-    } should have message singleError("#writable(all) is not compatible with #readable(this) in User")
+    } should have message {
```

Comment:
I will never stop being amazed at what you can do with this language

Reply:
scala best language